lib/plugins/openaps.js

```diff
@@ -23,17 +23,128 @@ function init() {
-      var openapsStatus = _.findLast(sbx.data.devicestatus, function (status) {
-        return sbx.entryMills(status) <= sbx.time && ('openaps' in status);
-      });
+    sbx.offerProperty('openaps', function setOpenAPS ( ) {
+      return openaps.analyzeData(sbx);
+    });
+  };
 
-      var loopStatus = checkLoopStatus(openapsStatus, sbx);
-      openapsStatus = openapsStatus || {};
-      openapsStatus.status = loopStatus;
+  openaps.analyzeData = function analyzeData (sbx) {
+    var recentHours = 6; //TODO dia*2
+    var recentMills = sbx.time - times.hours(recentHours).msecs;
 
-      return openapsStatus;
+    var recentData = _.filter(sbx.data.devicestatus, function (status) {
+      return ('openaps' in status) && sbx.entryMills(status) <= sbx.time && sbx.entryMills(status) >= recentMills;
     });
+
+    var prefs = openaps.getPrefs(sbx);
+    var recent = moment(sbx.time).subtract(prefs.warn / 2, 'minutes');
+
+    var result = {
+      seenDevices: {}
+      , lastEnacted: null
+      , lastNotEnacted: null
+      , lastSuggested: null
+      , lastIOB: null
+    };
+
+    function getDevice(status) {
+      var uri = status.device || 'device';
+      var device = result.seenDevices[uri];
+
+      if (!device) {
+        device = {
+          name: uri.indexOf('openaps://') === 0 ? uri.substring('openaps://'.length) : uri
+          , uri: uri
+        };
+
+        result.seenDevices[uri] = device;
+      }
+      return device;
+    }
+
+    function toMoments (status) {
+      return {
+        when:  moment(status.mills)
+        , enacted: status.openaps.enacted && status.openaps.enacted.recieved && moment(status.openaps.enacted.timestamp)
+        , notEnacted: status.openaps.enacted && !status.openaps.enacted.recieved && moment(status.openaps.enacted.timestamp)
+        , suggested: status.openaps.suggested && moment(status.openaps.suggested.timestamp)
+        , iob: status.openaps.iob && moment(status.openaps.iob.timestamp)
+      };
+    }
+
+    function momentsToLoopStatus (moments, noWarning)  {
+
+      var status = {
+        symbol: '⚠'
+        , code: 'warning'
+        , label: 'Warning'
+      };
+
+      if (moments.notEnacted && moments.notEnacted.isAfter(recent)) {
+        status.symbol = 'x';
+        status.code = 'notenacted';
+        status.label = 'Not Enacted';
+      } else if (moments.enacted && moments.enacted.isAfter(recent)) {
+        status.symbol = '⌁';
+        status.code = 'enacted';
+        status.label = 'Enacted';
+      } else if (moments.suggested && moments.suggested.isAfter(recent)) {
+        status.symbol = '↻';
+        status.code = 'looping';
+        status.label = 'Looping';
+      } else if (moments.when && (noWarning || moments.when.isAfter(recent))) {
+        status.symbol = '◉';
+        status.code = 'waiting';
+        status.label = 'Waiting';
+      }
+
+      return status;
+    }
+
+    _.forEach(recentData, function eachStatus (status) {
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [This function's cyclomatic complexity is too high. (11)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1941108539/issues/source?bid=2452379&fileBranchId=3088515#l104)